source/extensions/filters/http/cache/thundering_herd_handler.cc

```diff
@@ -0,0 +1,233 @@
+      ENVOY_STREAM_LOG(debug, "ThunderingHerdHandler adding blocker {}", *decoder_callbacks,
+                       request_headers.Path()->value().getStringView());
+      it->second.blockers_.push_back(maybeMakeTimeout(key, decoder_callbacks->dispatcher()));
+      lock.Release();
+      // Lock must be released before continueDecoding, because it can trigger filter
+      // destruction which can trigger another thundering herd action.
+      decoder_callbacks->continueDecoding();
+    } else {
+      ENVOY_STREAM_LOG(debug, "ThunderingHerdHandler adding blocked {}", *decoder_callbacks,
+                       request_headers.Path()->value().getStringView());
+      addBlocked(it->second, weak_filter, decoder_callbacks, request_headers);
+    }
+  }
+
+  void addBlocked(Queue& queue, std::weak_ptr<ThunderingHerdRetryInterface> weak_filter,
+                  Http::StreamDecoderFilterCallbacks* decoder_callbacks,
+                  Http::RequestHeaderMap& request_headers) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    queue.blocked_.push_back(
+        {weak_filter, decoder_callbacks, decoder_callbacks->dispatcher(), request_headers});
+  }
+
+  /**
+   * Returns nullptr if unblock_additional_request_period was not configured.
+   * Returns a timer that will expire after the configured duration if it was configured;
+   * the timer will release one item from the queue.
+   */
+  Event::TimerPtr maybeMakeTimeout(const Key& key, Event::Dispatcher& dispatcher) {
+    if (unblock_additional_request_period_ == std::chrono::milliseconds(0)) {
+      return {};
+    }
+    Event::TimerPtr timer = dispatcher.createTimer(
+        [handler = shared_from_this(), key]() { handler->selectNewBlocker(key); });
+    timer->enableTimer(unblock_additional_request_period_);
+    return timer;
+  }
+
+  std::function<void()> onUndeliverable(const Key& key) {
+    return [key, handler = shared_from_this()]() {
+      handler->handleInsertFinished(key, InsertResult::Failed);
+    };
+  }
+
+  void selectNewBlocker(const Key& key, Queue& queue) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    if (queue.blocked_.empty()) {
+      ENVOY_LOG(debug, "ThunderingHerdHandler ended blocker for {}", key.path());
+      if (queue.blockers_.empty()) {
+        ENVOY_LOG(debug, "ThunderingHerdHandler deleted queue for {}", key.path());
+        queues_.erase(key);
+      }
+      return;
+    }
+    Queue::Blocked unblocking = std::move(queue.blocked_.front());
+    queue.blocked_.pop_front();
+    queue.blockers_.push_back(maybeMakeTimeout(key, unblocking.dispatcher_));
+    unblocking.dispatcher_.post([unblocking, on_undeliverable = onUndeliverable(key)]() {
+      ENVOY_STREAM_LOG(debug, "ThunderingHerdHandler moving blocked to blocker",
+                       *unblocking.decoder_callbacks_);
+      auto filter = unblocking.weak_filter_.lock();
+      if (filter != nullptr) {
+        unblocking.decoder_callbacks_->continueDecoding();
+      } else {
+        on_undeliverable();
```

Comment:
"undeliverable" is ambiguous. I think something like "on_filter_deleted" would be more clear